go/vt/key/key.go

```diff
@@ -49,261 +52,258 @@ func (i Uint64Key) Bytes() []byte {
-// A hash vindex produces an 8-byte number.
-// This means that all numbers less than 0x8000000000000000 will fall in shard -80.
-// Any number with the highest bit set will be >= 0x8000000000000000, and will therefore
-// belong to shard 80-.
-// This means that from a keyrange perspective -80 == 00-80 == 0000-8000 == 000000-800000
-// If we don't add this padding, we could run into issues when transitioning from keyranges
-// that use 2 bytes to 4 bytes.
-func addPadding(kr []byte) []byte {
-	paddedKr := make([]byte, 8)
+// KeyRangeEndCompare compares the End of two KeyRange values using semantics unique to end values where an
+// empty End means the *maximum* value; returns -1 if a<b, 1 if a>b, 0 if equal.
+func KeyRangeEndCompare(a, b *topodatapb.KeyRange) int {
+	aIsMaximum := a == nil || Empty(a.End)
+	bIsMaximum := b == nil || Empty(b.End)
 
-	for i := 0; i < len(kr); i++ {
-		paddedKr = append(paddedKr, kr[i])
+	if aIsMaximum && bIsMaximum {
+		return 0
+	} else if aIsMaximum {
+		return 1
+	} else if bIsMaximum {
+		return -1
 	}
 
-	for i := len(kr); i < 8; i++ {
-		paddedKr = append(paddedKr, 0)
-	}
-	return paddedKr
+	return Compare(a.End, b.End)
 }
 
-// KeyRangeStartSmaller returns true if right's keyrange start is _after_ left's start
-func KeyRangeStartSmaller(left, right *topodatapb.KeyRange) bool {
-	if left == nil {
-		return right != nil
-	}
-	if right == nil {
-		return false
-	}
-	return bytes.Compare(left.Start, right.Start) < 0
+// KeyRangeEndEqual returns true if both KeyRange values have the same end.
+func KeyRangeEndEqual(a, b *topodatapb.KeyRange) bool {
+	return KeyRangeEndCompare(a, b) == 0
+}
+
+// KeyRangeCompare compares two KeyRange values, taking into account both the Start and End fields and their
+// field-specific comparison logic; returns -1 if a<b, 1 if a>b, 0 if equal. Specifically:
+//
+//   - The Start-specific KeyRangeStartCompare and End-specific KeyRangeEndCompare are used for proper comparison
+//     of an empty value for either Start or End.
+//   - The Start is compared first and End is only compared if Start is equal.
+func KeyRangeCompare(a, b *topodatapb.KeyRange) int {
+	// First, compare the Start field.
+	if v := KeyRangeStartCompare(a, b); v != 0 {
+		// The Start field for a and b differ, and that is enough; return that comparison.
+		return v
+	}
+
+	// The Start field was equal, so compare the End field and return that comparison.
+	return KeyRangeEndCompare(a, b)
+}
+
+// KeyRangeEqual returns true if a and b are equal.
+func KeyRangeEqual(a, b *topodatapb.KeyRange) bool {
+	return KeyRangeCompare(a, b) == 0
+}
+
+// KeyRangeLess returns true if a is less than b.
+func KeyRangeLess(a, b *topodatapb.KeyRange) bool {
+	return KeyRangeCompare(a, b) < 0
+}
+
+// KeyRangeIsComplete returns true if the KeyRange covers the entire keyspace.
+func KeyRangeIsComplete(keyRange *topodatapb.KeyRange) bool {
+	return keyRange == nil || (Empty(keyRange.Start) && Empty(keyRange.End))
```

Comment:
This is a good point, we should handle an `id` with any amount of zero bytes actually, and I think can be solved with a very small change: Empty should actually check `len(Normalize(id))` rather than `len(id)` – with my latest change to remove trailing zeroes in `Normalize()`, the length of an all-zero `id` would always be zero as well.